File: src/main/java/com/n26/service/serviceImpl/TransactionCacheImpl.java
package com.n26.service.serviceImpl;

import com.n26.model.Transaction;
import com.n26.service.TransactionCache;
import com.n26.service.serviceImpl.TransactionCacheHandler;
import org.springframework.cache.CacheManager;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;

import org.springframework.cache.caffeine.CaffeineCache;
import org.springframework.stereotype.Service;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.concurrent.ConcurrentMap;


@Service
public class TransactionCacheImpl implements TransactionCache {

    private CacheManager cacheManager;

    @Inject
    public TransactionCacheImpl(CacheManager cacheManager){
        this.cacheManager = cacheManager;
    }

    @Cacheable("TransactionCache")
    public Transaction cachingTransaction(Transaction transaction){
        return transaction;
    }

    public ArrayList getCacheValues(){
        CaffeineCache caffeineCache = (CaffeineCache)cacheManager.getCache("TransactionCache");
        ConcurrentMap<Object, Object> cache = caffeineCache.getNativeCache().asMap();
        return TransactionCacheHandler.getAmountsArrayList(cache);
    }



    @CacheEvict(value = "TransactionCache", allEntries = true)
    public void evictAllCacheValues() {}
}

File: src/main/kotlin/com/service/serviceImpl/TransactionCacheImpl.kt
package com.service.serviceImpl

import com.service.serviceImpl.TransactionCacheHandler.getAmountsList
import javax.inject.Inject
import org.springframework.cache.CacheManager
import com.service.TransactionCache
import org.springframework.cache.annotation.Cacheable
import com.model.Transaction
import mu.KotlinLogging
import java.math.BigDecimal
import org.springframework.cache.caffeine.CaffeineCache
import java.util.concurrent.ConcurrentMap
import org.springframework.cache.annotation.CacheEvict
import org.springframework.stereotype.Service

private val logger = KotlinLogging.logger {}

@Service
class TransactionCacheImpl @Inject constructor(private val cacheManager: CacheManager) : TransactionCache {
    @Cacheable("TransactionCache")
    override fun cachingTransaction(transaction: Transaction): Transaction {
        return transaction
    }

    override fun getCacheValues(): List<BigDecimal> {
        val caffeineCache = cacheManager.getCache("TransactionCache") as CaffeineCache
        val cache: ConcurrentMap<Any, Any> = caffeineCache.nativeCache.asMap()

        logger.debug { "getCacheValues return: ${getAmountsList(cache)}"}
        return getAmountsList(cache)
    }

    @CacheEvict(value = ["TransactionCache"], allEntries = true)
    override fun evictAllCacheValues() {
        logger.debug { "evictAllCacheValues"}
    }
}